### PR TITLE
Define patterns for the generation of MAC instructions

### PR DESCRIPTION
**Pull request by serkm**
Hello Mr Krones,

I want to inform you, that I implemented support for the ISA extension of the CV32E40P in LLVM as part of my bachelor thesis at the University of Tübingen, and I am now working on integration of my work in your project.

You can find my implementation here: https://github.com/ekut-es/pulp-llvm
My implementation includes definitions for the Instructions of the ISA extension, code generation for these instructions, generation of hardware loops and builtin functions for the direct use of these instructions.

### CODE REVIEW

**Comment by flip1995** [comment]
Please call me Phil, Philipp or just by my GH handle @flip1995  :)

This is great, thanks for the contribution! I try to review this soon.

I saw that you also implemented other extensions like the ALU operations. Any chance, that you're motivated in porting those to this repo? If you want to get more involved with this work, I can recommend joining the [OpenHW Group Mattermost](https://mattermost.openhwgroup.org/all-users/channels/twg--sw--llvm-tools).

For us to merge this, you'll have to sign the [ECA](https://www.eclipse.org/legal/ECA.php). If you already did this, make sure that the email used for committing is the same as in the agreement. Finally, you have to sign your commit, which you can fix with `git commit --amend --signoff && git push -f`.

**Comment by serkm** [comment]
Yes, I'll port all the instruction definitions and the code generation.
I'll try to stick to your code style and add tests wherever needed.

**Comment by flip1995** [comment]
Great, thanks! You may also be interested in joining the monthly Zoom meeting of the Software TG of the OpenHW group. The next call will be on Monday 2021-02-08 4PM CET. You can find more information on the [OpenHW Group Calendar](https://www.openhwgroup.org/working-groups/) (at the bottom of the page). Here's a [direct link](https://www.google.com/calendar/event?eid=Mm1tZjJvNXBpcGRoY2U1OHZiajFkOWNrdTRfMjAyMTAzMDhUMTUwMDAwWiBtZWV0aW5nc0BvcGVuaHdncm91cC5vcmc&ctz=America/Toronto) to the event, which may or may not work.

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 1]
```diff
@@ -175,3 +175,99 @@ let Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore =
                                  "cv.mulhhurn", "$rd, $rs1, $rs2, $imm5", []>,
                     Sched<[]>;
 } // Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore = 0
+
+
+// Helper fragments
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]:
```suggestion
//===----------------------------------------------------------------------===//
// CORE-V specific helper fragments
//===----------------------------------------------------------------------===//
```

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 2]
```diff
@@ -175,3 +175,99 @@ let Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore =
...
+
+def ushiftRound : PatFrag<(ops node:$value, node:$shiftAmount),
+                          (srl (add node:$value, powerOf2), node:$shiftAmount), [{
+
+  if (auto powerOf2 = dyn_cast<ConstantSDNode>(N->getOperand(0)->getOperand(1)))
+    return (powerOf2->getZExtValue() << 1) == (1U << N->getConstantOperandVal(1));
+  return false;
+}]>;
+
+// Patterns for mac operations
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]:
```suggestion
//===----------------------------------------------------------------------===//
// Patterns for MAC operations
//===----------------------------------------------------------------------===//
```

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 3]
```diff
@@ -175,3 +175,99 @@ let Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore =
...
+
+def muls   : PatFrag<(ops node:$rs1, node:$rs2),
+                     (mul (sext_inreg node:$rs1, i16), (sext_inreg node:$rs2, i16))>;
+def mulhhs : PatFrag<(ops node:$rs1, node:$rs2),
+                     (mul (sra node:$rs1, (i32 16)), (sra node:$rs2, (i32 16)))>;
+def mulu   : PatFrag<(ops node:$rs1, node:$rs2),
+                     (mul (and node:$rs1, 65535), (and node:$rs2, 65535))>;
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]:
```suggestion
                     (mul (and node:$rs1, 0xffff), (and node:$rs2, 0xffff))>;
```

File: llvm/test/CodeGen/RISCV/corev/mac.ll [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+define i32 @msu(i32 %a, i32 %b, i32 %c) {
+; CHECK-LABEL: msu:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    mul a0, a0, a1
+; CHECK-NEXT:    sub a0, a0, a2
+; CHECK-NEXT:    ret
+  %1 = mul i32 %a, %b
+  %2 = sub i32 %1, %c
```
Comment (flip1995) [inline llvm/test/CodeGen/RISCV/corev/mac.ll]:
```suggestion
  %2 = sub i32 %c, %1
```

With the operands swapped this should produce `cv.msu` I think.

**Review by flip1995 — COMMENTED** [review]
LGTM overall. The `msu` test needs a fix. Otherwise only some formatting for consistency.

**Review by flip1995 — APPROVED** [review]
LGTM @jeremybennett This can be merged. Tests pass for me locally.

**Comment by jeremybennett** [comment]
Thanks for the contribution. All merged.

**Pull request merged**